### DISCUSSION TITLE
Adding to coroutine context in V4.x.x

### DISCUSSION THREAD

**Discussion by omriperl** [ExpediaGroup/graphql-kotlin#1777] [Q&A]
Hello,

I'm looking for a solution to update the coroutine context with my custom elements (e.g., MDCContext) in version 4.x.x of the Expedia GraphQL Kotlin library. I have tried implementing my own GraphQLRequestHandler and returning the result within a custom coroutine scope and context, but it hasn't been successful.

I noticed that in version 6.x.x of the library, there is documentation (https://opensource.expediagroup.com/graphql-kotlin/docs/server/graphql-context-factory) mentioning the ability to update the context, but I'm specifically interested in achieving this in version 4.x.x.

I would greatly appreciate any assistance or guidance on this matter. Thank you!

**Comment by dariuszkuc** [comment]
Hello 👋 
You are halfway there -> in order to get context propagation working you would need to 1) create custom request handler (or server) that populates coroutine context information in the GraphQL context and 2) create custom function data fetcher that uses that scope to run your functions ([link](https://github.com/ExpediaGroup/graphql-kotlin/blob/6.x.x/generator/graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/execution/FunctionDataFetcher.kt#L117)).

One caveat here is that v4 is based on `graphql-java` v16 which still used arbitrary Object as context (so you have to define your context instance to have the coroutine scope field). Generic context map mechanism which simplifies this was introduced to `graphql-java` in v17.

I'd highly recommend to upgrade to newer versions of the lib which includes number of improvements and potentially fixes some security issues with old libs.

Thanks,
Derek

**Reply by omriperl** [reply]
Hey @dariuszkuc 
I will try that, thanks!

**Reply by omrip30** [reply]
Working great @dariuszkuc  ! Really appreciated :)
